File: liftingline.m
% [CL,CD,Y,GAMMA,V] = LIFTINGLINE(GEOMFILE,FORCEFILE) ...
% [CL,CD,Y,GAMMA,V] = LIFTINGLINE(GEOMFILE,FORCEFILE,ALFAS) ...
% [CL,CD,Y,GAMMA,V] = LIFTINGLINE(GEOMFILE,FORCEFILE,ALFAS,'relfactor',R)
% 
% Prandtl's Lifting-Line Theory (LLT) solver for a finite wing.
% See Anderson, Fundamentals of Aerodynamics, 5th ed., Ch. 5.
% 
% INPUTS
% ------
% GEOMFILE is the geometry file for the wing, with columns for
% spanwise position and local chord length (both normalized by total span), 
% and local twist angles (in degrees).
% 
% FORCEFILE contains the airfoil lift and drag data, with columns for angle
% of attack (deg), lift coefficient and drag coefficient (in that order).
% Additional columns will be ignored.
% 
% ALFAS is an optional third argument containing the wing angles of attack
% (deg) at which to perform the calculation.  It should be a horizontal 
% vector. If ALFAS is not specified, a wing angle of attack of zero is
% assumed.
% 
% 'relfactor' is an optional name-value pair controlling relaxation (default
% 0.01).  Smaller values improve stability but require more iterations.
%
% OUTPUTS
% -------
% CL and CD are the calculated lift and induced drag coefficients on the 
% full wing.  They will have the same dimensions as ALFAS.
% 
% Y is the spanwise coordinate vector extracted from GEOMFILE,
% normalized by total span.
% 
% GAMMA holds the spanwise circulation distribution for each angle of
% attack in a column pertaining to the position of that angle of attack in
% the ALFAS vector.  Circulation is normalized by freestream velocity and
% total span.
% 
% V holds the spanwise downwash distribution (normalized by freestream 
% velocity) in columns corresponding to each angle of attack in ALFAS, as 
% with GAMMA. 
%
% EXAMPLE
% -------
% [CL, CD] = liftingline('elliptic_AR8.txt', 'thin_airfoil_data.txt', ...
%                        -10:2:10, 'relfactor', 0.01);
% plot(-10:2:10, CL)
%
% The Python equivalent is available at limacher-llt/liftingline.py.
%
% REFERENCES
% ----------
% Anderson, J. D. Jr. (2011). Fundamentals of Aerodynamics, 5th ed.
%   McGraw-Hill.  See Ch. 5 (pp. 449-470) for lifting-line theory,
%   Eq. (5.69) for elliptic-wing lift slope, and Eq. (5.61) for induced drag.
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


function [CL,CD,y,Gamma,v] = liftingline(geomfile,forcefile,varargin)

    % Default values
    AoA = 0;
    relfactor = 0.01;

    % Read optional input of angles of attack
    if nargin > 2
        if ~ischar(varargin{1})
            AoA = varargin{1};
        end
        for ii = 1:(nargin - 2)
            if ischar(varargin{ii})
                switch varargin{ii}
                    case 'relfactor'
                        relfactor = varargin{ii+1};
                end
            end
        end   
    end

    % Ensure AoA is a horizontal vector
    [tempa,tempb] = size(AoA);
    if tempa > tempb
        AoA = transpose(AoA);
    elseif tempa ~= 1 && tempb ~= 1
        disp('AoA cannot be a matrix.');
        return;
    end
    
    % Read force data
        data = dlmread(forcefile, ',', 1, 0);
    alfaref = data(:,1);
    cl = data(:,2);
    cd = data(:,3);

    % Read blade geometry data
        data = dlmread(geomfile, ',', 1, 0);
    y = data(:,1);
    c = data(:,2); 
    th = data(:,3);

    % Ensure chord and span coordinates are normalized by total span
    L = y(end)-y(1); 
    y = y/L;
    c = c/L;

    %%%%%%%%%%%%%%%%%%
    % Check that lift curve is monotonic
    [out, monoInds] = checkmonotonic(cl);
    if out == 0
        disp('Lift coefficient data is not monotonically increasing.')
        disp('  Data will be truncated to its longest monotonic section.')
    end

    % Truncate forcefile to include only monotonic lift section
    alfaref = alfaref(monoInds);
    cl = cl(monoInds);
    cd = cd(monoInds);
    %%%%%%%%%%%%%%%%%%%%%%

    % Intialize solution variables
    temp = zeros(length(y),length(AoA));
    Gamma = temp; 
    v = temp;
    % alfai = temp;
    CL = zeros(size(AoA));
    CD = zeros(size(AoA));

    % Solve
    for ii = 1:length(AoA)
        
        %%%%%%%%%%%%%%%%%
        % [CL(ii),CD(ii),Gamma(:,ii),v(:,ii)] = ...
        %     LLsingle(y,c,th,alfaref,cl,AoA(ii),relfactor);
        [CL(ii),CD(ii),Gamma(:,ii),v(:,ii)] = ...
            LLsingle(y,c,th,alfaref,cl,cd,AoA(ii),relfactor);
        %%%%%%%%%%%%%%%%%

        
    end

end

%%%%%%%%%%
function [CL,CD,Gamma,v] = LLsingle(y,c,th,alfaref,cl,cd,AoA,relfactor)
    
    % Iterate to solve for circulation distribution
    [Gamma,v,alfai,~] = calcgamma(y,c,th,AoA,alfaref,cl,...
        'relfactor',relfactor);

    % Calculate lift and drag
    cli = interp1(alfaref,cl,alfai,'linear','extrap');
    
    %%%%%%%%%%%%%%%%
    % cdi = zeros(size(cli)); % high-Re assumption; flow attached, no skin friction
    cdi = interp1(alfaref,cd,alfai,'linear','extrap');
    %%%%%%%%%%%%%%%%

    % Ensure force coefficients go to zero at ends
    cli(1) = 0; cdi(1) = 0; cli(end) = 0; cdi(end) = 0;
    
    % Calculate planform area
    S = trapz(y,c);  

    % Calculate 
    CL = (1/S) * trapz(y, c.*(cli.*cos(v) - cdi.*sin(v)));
    CD = (1/S) * trapz(y, c.*(cli.*sin(v) + cdi.*cos(v)));

end


%%%%%%%%%%%%%

function [Gammai,vi,alfai,varargout] = calcgamma(yi,ci,thi,AoA,alfa,cl,varargin)

    % Prepare for optional output
    if nargout > 3
        varargout{1} = [];
    end

    % Default settings
    Gamma0 = 0;         % scaling of initial guess for Gamma
    maxiter = 1000;     % maximum iterations
    errtol = 1e-6;      % error tolerance
    relfactor = 0.01;   % relaxation factor    

    % Optional input arguments
    if ~isempty(varargin)
        opts = varargin;
        for ii = 1:length(opts)
            switch opts{ii}
                case 'Gamma0'
                    Gamma0 = opts{ii+1};
                case 'maxiter' 
                    maxiter = opts{ii+1};
                case 'errtol'
                    errtol = opts{ii+1};
                case 'relfactor'
                    relfactor = opts{ii+1};
            end
        end
    end

    % Convert angles to radians
    alfa = pi/180 * alfa;
    AoA = pi/180 * AoA;
    thi = pi/180 * thi;

    % Circulation to be calculated at locations between those in y vector
    yj = 0.5*(yi + circshift(yi,-1));
    yj = yj(1:end-1);

    % Initialize variables
    alfai = zeros(size(yi));
    Gammai = zeros(size(yi));
    cli = zeros(size(yi));

    % Intial guess of Gamma distribution, defined on yi
    Gammai = -Gamma0*(yi-yi(1)).*(yi-yi(end)); % parabolic distribution, zero at ends

    % Ensure end conditions are met
    Gammai(1) = 0;
    Gammai(end) = 0;

    % Iterate until Gamma is converged
    err = 10;
    iter = 0;
    while iter < maxiter && err(end) > errtol
    
        iter = iter + 1;

        % Trailing vortex strengths
        dGammaj = -diff(Gammai);
        
        % Calculate influence matrix
        Yij = 1./( 4* pi * ( yj' - yi ) );
        
        % Calculate induced velocities
        vi = Yij * dGammaj;
        
        % Calculate effective angle of attack
        alfai = AoA + thi - vi;
        %%%%%%%%%%%%%%
        % cli =  interp1(alfa,cl,alfai);
        cli =  interp1(alfa,cl,alfai,'linear','extrap');
        %%%%%%%%%%%%%
        cli(1) = 0; cli(end) = 0;

        % Calculate circulation from lift coefficient
        Gammai_new = cli .* ci/2;
        Gammai_new(1) = 0;
        Gammai_new(end) = 0;

        % Calculate error
        eGamma = Gammai_new - Gammai;
        err(end+1) = max(abs(eGamma));

        % Iterate on Gamma
        Gammai = Gammai + relfactor * eGamma;
    end  

    % Convergence report
    disp(['Total iterations: ',num2str(iter)])

    disp(['err = ', num2str(err(end))])
    if err(end) > errtol
        disp(['NOT CONVERGED WITHIN e = ',num2str(errtol)])
    else
        disp(['CONVERGED WITHIN e = ',num2str(errtol)])
    end
    %%%%%%%%%%
        % Check if angle of attack is out of range
        if any(alfai(2:end-1)>alfa(end)) | any(alfai(2:end-1)<alfa(1))
            disp('ERROR: alfa out of range');
            Gammai = NaN;
            vi = NaN;
            return;
        end
%%%%%%%%%%

    if nargout > 3
        varargout{1} = err;
    end

    % Warning
    % if any(alfai(2:end-1) > alfa(end)) || ...
    %         any(alfai(2:end-1) < alfa(1))
    % 
    %     warning(['Converged effective angle of attack lies outside ', ...
    %          'the supplied airfoil dataset.']);
    % end

    % Convert back to degrees
    alfai = 180/pi * alfai;

end

% function out = checkmonotonic(x)
% 
%     temp = sum(diff(x)<=0);
% 
%     if temp ~= 0
%         disp('ERROR:');
%         out = 0;
%     else
%         out = 1;
%     end
% 
% end

%%%%%%%%%%%%%%%%%
function [flag, indices] = checkmonotonic(x)
%CHECKMONOTONIC Check whether a vector is strictly monotonically increasing.
%
%   [flag, indices] = checkmonotonic(x)
%
%   flag    = 1 if the entire vector is strictly increasing
%             0 otherwise
%
%   indices = indices of the longest contiguous strictly increasing section,
%             with the same row or column orientation as x

    flag = all(diff(x) > 0);

    if flag
        indices = 1:numel(x);
    else
        breaks = find(diff(x) <= 0);

        starts = [1; breaks(:) + 1];
        ends   = [breaks(:); numel(x)];

        [~, iLongest] = max(ends - starts + 1);

        indices = starts(iLongest):ends(iLongest);
    end

    % Match the orientation of x
    if iscolumn(x)
        indices = indices(:);
    end
end
%%%%%%%%%%%%%%%%%


%%%%%%%%%%%%%%%%

File: liftingline.py
"""
liftingline.py — Prandtl Lifting-Line Theory solver

Port of the MATLAB implementation (limacher-llt/liftingline.m) to Python.

Functions
---------
liftingline(geomfile, forcefile, AoA=0, relfactor=0.01)
    Main lifting-line solver. Returns CL, CD, y, Gamma, v.

LLsingle(y, c, th, alfaref, cl, cd, AoA, relfactor)
    Solve for a single angle of attack.

calcgamma(yi, ci, thi, AoA, alfa, cl, ...)
    Iterative circulation solver with relaxation.

checkmonotonic(x)
    Check if a vector is strictly monotonically increasing.
"""

import warnings

import numpy as np


# ═════════════════════════════════════════════════════════════════════════════
#  Main function
# ═════════════════════════════════════════════════════════════════════════════

def liftingline(geomfile, forcefile, AoA=0, relfactor=0.01):
    """
    [CL, CD, y, Gamma, v] = liftingline(geomfile, forcefile, AoA)

    Parameters
    ----------
    geomfile : str
        Geometry file with columns: spanwise position (normalised by span),
        chord length (normalised by span), twist angle (deg).
    forcefile : str
        Airfoil data file with columns: angle of attack (deg), lift
        coefficient, drag coefficient.  Additional columns are ignored.
    AoA : float or array_like, optional
        Wing angles of attack (deg).  Scalar or 1-D array.  Default 0.

    Returns
    -------
    CL : ndarray
        Lift coefficient for each angle in AoA.
    CD : ndarray
        Induced drag coefficient for each angle in AoA.
    y : ndarray
        Spanwise stations normalised by span.
    Gamma : ndarray
        Normalised circulation  (rows = spanwise stations,
        columns = angles of attack).
    v : ndarray
        Normalised downwash, v/U  (same shape as Gamma).
    """
    # Ensure AoA is a 1-D array
    AoA = np.atleast_1d(np.asarray(AoA, dtype=float)).ravel()

    # Read force data
    data = np.loadtxt(forcefile)
    alfaref = data[:, 0]
    cl = data[:, 1]
    cd = data[:, 2]

    # Read blade geometry data
    data = np.loadtxt(geomfile, skiprows=1)          # header row
    y = data[:, 0]
    c = data[:, 1]
    th = data[:, 2]

    # Validate geometry input
    N = len(y)
    if N < 2:
        raise ValueError(
            'Geometry file must contain at least 2 spanwise stations.'
        )
    if np.any(np.diff(y) <= 0):
        raise ValueError(
            'Spanwise coordinates must be strictly increasing.'
        )

    # Normalise chord and span coordinates by total span
    L = y[-1] - y[0]
    y = y / L
    c = c / L

    # Check that the lift curve is monotonically increasing
    _, monoInds = checkmonotonic(cl)
    if len(monoInds) < len(cl):
        print('Lift coefficient data is not monotonically increasing.')
        print('  Data will be truncated to its longest monotonic section.')

    alfaref = alfaref[monoInds]
    cl = cl[monoInds]
    cd = cd[monoInds]

    # Initialise solution arrays
    n_y = len(y)
    n_aoa = len(AoA)
    Gamma = np.zeros((n_y, n_aoa))
    v = np.zeros((n_y, n_aoa))
    CL = np.zeros(n_aoa)
    CD = np.zeros(n_aoa)

    # Solve for each angle of attack
    for ii, aoa in enumerate(AoA):
        CL[ii], CD[ii], Gamma[:, ii], v[:, ii] = (
            LLsingle(y, c, th, alfaref, cl, cd, aoa, relfactor)
        )

    return CL, CD, y, Gamma, v


# ═════════════════════════════════════════════════════════════════════════════
#  Single-angle solver
# ═════════════════════════════════════════════════════════════════════════════

def LLsingle(y, c, th, alfaref, cl, cd, AoA, relfactor):
    """
    Solve lifting-line for a single angle of attack.

    Parameters
    ----------
    y : ndarray       Normalised spanwise stations.
    c : ndarray       Normalised chord distribution.
    th : ndarray       Twist angles (deg).
    alfaref : ndarray  Reference angles of attack from airfoil data (deg).
    cl : ndarray       Lift coefficient data.
    cd : ndarray       Drag coefficient data.
    AoA : float        Wing angle of attack (deg).

    Returns
    -------
    CL : float
    CD : float
    Gamma : ndarray    Normalised circulation.
    v : ndarray        Normalised downwash (v/U).
    """
    # Iterative circulation solver
    Gamma, v, alfai, _ = calcgamma(
        y, c, th, AoA, alfaref, cl, relfactor=relfactor
    )

    # Interpolate force coefficients at converged effective AoA
    cli = np.interp(alfai, alfaref, cl)
    cdi = np.interp(alfai, alfaref, cd)

    # Force coefficients go to zero at blade ends
    cli[0] = 0.0
    cdi[0] = 0.0
    cli[-1] = 0.0
    cdi[-1] = 0.0

    # Planform area (dimensionless)
    S = np.trapz(c, y)

    # Total force coefficients (eqs. 17-18 in LLT.tex)
    CL = (1.0 / S) * np.trapz(
        c * (cli * np.cos(v) - cdi * np.sin(v)), y
    )
    CD = (1.0 / S) * np.trapz(
        c * (cli * np.sin(v) + cdi * np.cos(v)), y
    )

    return CL, CD, Gamma, v


# ═════════════════════════════════════════════════════════════════════════════
#  Iterative circulation solver
# ═════════════════════════════════════════════════════════════════════════════

def calcgamma(yi, ci, thi, AoA, alfa, cl,
              Gamma0=0, maxiter=1000, errtol=1e-6, relfactor=0.01):
    """
    Iteratively solve for the circulation distribution.

    Parameters
    ----------
    yi : ndarray       Normalised spanwise stations.
    ci : ndarray       Normalised chord.
    thi : ndarray       Twist angles (deg).
    AoA : float        Wing angle of attack (deg).
    alfa : ndarray     Reference angles of attack (deg).
    cl : ndarray       Lift coefficient data.

    Keyword Args
    ------------
    Gamma0 : float     Scaling factor for initial parabolic guess.
    maxiter : int      Maximum iterations.
    errtol : float     Convergence tolerance on Gamma.
    relfactor : float  Relaxation factor (0 < r ≤ 1).

    Returns
    -------
    Gamma : ndarray    Converged normalised circulation.
    vi : ndarray       Normalised downwash (v/U).
    alfai : ndarray    Converged effective angle of attack (deg).
    err : ndarray      Error history.
    """
    # Convert angles to radians
    alfa_rad = np.deg2rad(alfa)
    AoA_rad = np.deg2rad(AoA)
    thi_rad = np.deg2rad(thi)

    # Trailing vortex positions at midpoints between yi stations
    yj = 0.5 * (yi + np.roll(yi, -1))[:-1]

    # Initial guess: parabolic distribution, zero at ends
    Gamma = -Gamma0 * (yi - yi[0]) * (yi - yi[-1])
    Gamma[0] = 0.0
    Gamma[-1] = 0.0

    # Iterate
    err = 10.0
    history = [err]
    n_iter = 0

    while n_iter < maxiter and err > errtol:
        n_iter += 1

        # Trailing vortex strengths (Helmholtz)
        dGamma = -np.diff(Gamma)

        # Influence matrix: Y[i, j] = 1 / (4 pi (yj[j] - yi[i]))
        Yij = 1.0 / (4.0 * np.pi * (yj[np.newaxis, :] - yi[:, np.newaxis]))

        # Induced downwash
        vi = Yij @ dGamma

        # Effective angle of attack (radians)
        alfai_rad = AoA_rad + thi_rad - vi

        # Interpolate lift coefficient (extrapolate during iteration)
        cli = np.interp(np.rad2deg(alfai_rad), alfa, cl)
        cli[0] = 0.0
        cli[-1] = 0.0

        # Circulation from lift coefficient  (eq. 14 in LLT.tex)
        Gamma_new = cli * ci / 2.0
        Gamma_new[0] = 0.0
        Gamma_new[-1] = 0.0

        # Error and relaxation update
        eGamma = Gamma_new - Gamma
        err = np.max(np.abs(eGamma))
        history.append(err)

        Gamma = Gamma + relfactor * eGamma

    # Convergence report
    print(f'Total iterations: {n_iter}')
    print(f'err = {err:.6e}')
    if err > errtol:
        print(f'NOT CONVERGED WITHIN e = {errtol}')
    else:
        print(f'CONVERGED WITHIN e = {errtol}')

    # Effective angle of attack at convergence (radians → degrees)
    alfai = np.rad2deg(alfai_rad)

    # Check if converged effective AoA is within supplied airfoil data range
    interior = alfai[1:-1]
    if np.any(interior > alfa[-1]) or np.any(interior < alfa[0]):
        print('ERROR: Effective angle of attack out of range.')
        Gamma[:] = np.nan
        vi[:] = np.nan
        return Gamma, vi, alfai, history

    return Gamma, vi, alfai, history


# ═════════════════════════════════════════════════════════════════════════════
#  Utility
# ═════════════════════════════════════════════════════════════════════════════

def checkmonotonic(x):
    """
    Check whether a vector is strictly monotonically increasing.

    Parameters
    ----------
    x : array_like

    Returns
    -------
    flag : bool
        True if the entire vector is strictly increasing.
    indices : ndarray
        Indices of the longest contiguous strictly increasing section
        (1-D, matches orientation of `x`).
    """
    x = np.asarray(x)
    flag = bool(np.all(np.diff(x) > 0))

    if flag:
        indices = np.arange(len(x))
    else:
        breaks = np.where(np.diff(x) <= 0)[0]
        starts = np.concatenate([[0], breaks + 1])
        ends = np.concatenate([breaks, [len(x) - 1]])
        longest = np.argmax(ends - starts + 1)
        indices = np.arange(starts[longest], ends[longest] + 1)

    # Match orientation of input
    if isinstance(x, np.matrix) or (hasattr(x, 'shape') and x.ndim == 2
                                     and x.shape[0] > x.shape[1]):
        indices = indices[:, np.newaxis]

    return flag, indices